security/advancedtls/crl_provider_test.go

```diff
@@ -0,0 +1,350 @@
+	}
+	if regularProvider.opts.RefreshDuration != 2*time.Hour {
+		t.Fatalf("Valid refreshDuration was incorrectly updated by validate() func")
+	}
+	regularProvider.Close()
+}
+
+// TestFileWatcherCRLProvider tests how FileWatcherCRLProvider handles the major
+// four cases for CRL checks. It scans the CRLs under crl directory to populate
+// the in-memory storage. Then we construct unrevoked, revoked leaf, and revoked
+// intermediate chains, as well as a chain without CRL for issuer, and check
+// that it’s correctly processed. Additionally, we also check if number of
+// invocations of custom callback is correct.
+func (s) TestFileWatcherCRLProvider(t *testing.T) {
+	const nonCRLFilesUnderCRLDirectory = 15
+	nonCRLFilesSet := make(map[string]struct{})
+	customCallback := func(err error) {
+		nonCRLFilesSet[err.Error()] = struct{}{}
+	}
+	p, err := NewFileWatcherCRLProvider(FileWatcherOptions{
+		CRLDirectory:               testdata.Path("crl"),
+		RefreshDuration:            1 * time.Hour,
+		CRLReloadingFailedCallback: customCallback,
+	})
+	if err != nil {
+		t.Fatal("Unexpected error while creating FileWatcherCRLProvider:", err)
+	}
+	p.Close()
+
+	// Each test data entry contains a description of a certificate chain,
+	// certificate chain itself, and if CRL is not expected to be found.
+	tests := []struct {
+		desc        string
+		certs       []*x509.Certificate
+		expectNoCRL bool
+	}{
+		{
+			desc:  "Unrevoked chain",
+			certs: makeChain(t, testdata.Path("crl/unrevoked.pem")),
+		},
+		{
+			desc:  "Revoked Intermediate chain",
+			certs: makeChain(t, testdata.Path("crl/revokedInt.pem")),
+		},
+		{
+			desc:  "Revoked leaf chain",
+			certs: makeChain(t, testdata.Path("crl/revokedLeaf.pem")),
+		},
+		{
+			desc:        "Chain with no CRL for issuer",
+			certs:       makeChain(t, testdata.Path("client_cert_1.pem")),
+			expectNoCRL: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			for _, c := range tt.certs {
+				crl, err := p.CRL(c)
+				if err != nil {
+					t.Fatalf("Expected error fetch from provider: %v", err)
+				}
+				if crl == nil && !tt.expectNoCRL {
+					t.Fatalf("CRL is unexpectedly nil")
+				}
+			}
+		})
+	}
+	if diff := cmp.Diff(len(nonCRLFilesSet), nonCRLFilesUnderCRLDirectory); diff != "" {
+		t.Errorf("Unexpected number Number of callback executions\ndiff (-got +want):\n%s", diff)
+	}
+}
+
+// TestFileWatcherCRLProviderDirectoryScan tests how FileWatcherCRLProvider
+// handles different contents of FileWatcherOptions.CRLDirectory.
+// We update the content with various (correct and incorrect) CRL files and
+// check if in-memory storage was properly updated. Please note that the same
+// instance of FileWatcherCRLProvider is used for the whole test so test cases
+// are not independent from each other.
+func (s) TestFileWatcherCRLProviderDirectoryScan(t *testing.T) {
+	sourcePath := testdata.Path("crl")
+	targetPath := createTmpDir(t)
```

Comment:
Shouldn't we remove this directory once the tests are finished? Or are all the Go versions we support recent enough for us to use https://godoc.corp.google.com/pkg/testing#T.TempDir?

Reply:
Good point, added the 'remove dir' approach.